flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/TestSchemaResolver.java

```diff
@@ -0,0 +1,199 @@
+                resolvedColumns[i] =
+                        resolveMetadataColumn((Schema.UnresolvedMetadataColumn) unresolvedColumn);
+            } else if (unresolvedColumn instanceof Schema.UnresolvedComputedColumn) {
+                resolvedColumns[i] =
+                        Column.computed(
+                                        unresolvedColumn.getName(),
+                                        resolveExpression(
+                                                ((Schema.UnresolvedComputedColumn) unresolvedColumn)
+                                                        .getExpression()))
+                                .withComment(unresolvedColumn.getComment().orElse(null));
+            }
+            i++;
+        }
+        return Arrays.asList(resolvedColumns);
+    }
+
+    private Column.PhysicalColumn resolvePhysicalColumn(
+            Schema.UnresolvedPhysicalColumn unresolvedColumn) {
+        return Column.physical(
+                        unresolvedColumn.getName(),
+                        dataTypeFactory.createDataType(unresolvedColumn.getDataType()))
+                .withComment(unresolvedColumn.getComment().orElse(null));
+    }
+
+    private Column.MetadataColumn resolveMetadataColumn(
+            Schema.UnresolvedMetadataColumn unresolvedColumn) {
+        return Column.metadata(
+                        unresolvedColumn.getName(),
+                        dataTypeFactory.createDataType(unresolvedColumn.getDataType()),
+                        unresolvedColumn.getMetadataKey(),
+                        unresolvedColumn.isVirtual())
+                .withComment(unresolvedColumn.getComment().orElse(null));
+    }
+
+    private WatermarkSpec resolveWatermarkSpecs(
+            Schema.UnresolvedWatermarkSpec unresolvedWatermarkSpec) {
+        return WatermarkSpec.of(
+                unresolvedWatermarkSpec.getColumnName(),
+                resolveExpression(unresolvedWatermarkSpec.getWatermarkExpression()));
+    }
+
+    private @Nullable UniqueConstraint resolvePrimaryKey(
+            @Nullable Schema.UnresolvedPrimaryKey unresolvedPrimaryKey) {
+        if (unresolvedPrimaryKey == null) {
+            return null;
+        }
+
+        final UniqueConstraint primaryKey =
+                UniqueConstraint.primaryKey(
+                        unresolvedPrimaryKey.getConstraintName(),
+                        unresolvedPrimaryKey.getColumnNames());
+
+        return primaryKey;
+    }
+
+    private ResolvedExpression resolveExpression(Expression expression) {
+        if (expression instanceof SqlCallExpression) {
+            String callString = ((SqlCallExpression) expression).getSqlExpression();
+            if (resolveExpressionTable.containsKey(callString)) {
+                return resolveExpressionTable.get(callString);
+            }
+        }
+        throw new IllegalArgumentException("Unsupported expression: " + expression);
+    }
+
+    public TestSchemaResolver addExpression(
```

Comment:
Do we really need this method?
TBH, it seems a little hard to understand the logic to resolve `UnresolvedComputedColumn`.

I think in the newly introduced `TestSchemaResolver`, it's not necessary to support  `UnresolvedComputedColumn`  in this PR. if someday we need, we can then support it in other pr.
If we decide to not support `UnresolvedComputedColumn`, please comment it in the java doc of this class.

Reply:
Actually, it's needed. You can refer to `HiveCatalogGenericMetadataTest`. We need this to work with test with computed columns.

Reply:
What I mean is do we really need it in this PR. As far as I see, before this pr, there is no test will need method `addExpression`. Without it, we won't break any existing tests.
If we need it in some day, we  can add it then,

Reply:
> As far as I see, before this pr, there is no test will need method addExpression

That is due to it compare `TableSchema` with `TableSchema` before. So for the watermark spec or computed columns, it was compared in the string format. But, now it should compared in the `ResolvedSchema` format (for unified expression and resolved dataType). So, we need a resolved expression for comparing.

Reply:
But did the origin test contain any resolved expression while creating TableSchema?

Reply:
No, before we compare the `TableSchema` to `TableSchema`, so the expression are all not resolved. Now, we use `ResolvedSchema` to create table, but get `Schema` when `getTable`. 
Comparing `Schema` is not feasible/easy. So we add the `TestSchemaResolver` to help compare the resolved schema. Then resolve expression ability is needed

Reply:
I understand the resolve expression ability is needed. But just consider this migrate work, I don't think we need to cover this since it won't break currernt testing without resolving expression ability TestSchemaResolver.
If we do need it, I think it'll be better to do in other pr.
TBH, I'm not intended to add code to cover extra test logic for the migration work.

Reply:
IMO, due to it's a migration work. So it can't do the exactly same this as before. for example:

Before we compare the `TableSchema` with `TableSchema`. In `TableSchema`, the dataTypes are resolved, but the expressions are unresolved.
Now, we have two choice: compare `Schema` with `Schema` or `ResolvedSchema` with `ResolvedSchema`. Both are not exactly same as before. Because the new framework split out the resolved and unresolved part. 

So from this point, we must have to adapt the test to the new api. So we have two way to do this:

- Compare `ResolvedSchema` with `ResolvedSchema`, it need the ability of resolve, so we introduce the `TestSchemaResolver`
- Or compare `Schema` with `Schema`. Due to the `Schema` is not comparable. So we have to write the expected `Schema` manually and `ResolvedSchema` for creating table. Then compare the string representation of the `getTable`'s  result and the expected `Schema`.

Both is acceptable for me, but the first one looks more convenient and less boilerplate code, So I use the first choice.

Reply:
FYI, the computed column and watermark spec is in original test, you can see `org.apache.flink.table.catalog.hive.HiveCatalogGenericMetadataTest#testGenericTableSchema`